## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [30]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone                                email  client_id  order_id  order_date  order_week  order_year         item_id     category         subcategory  unit_price  unit_cost  unit_weight  qty  line_number
0     Donald   Harding           Immunologist  793-904-7725x39308     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28          17        2023  EUD29711-63-6U        decor            wall art     1096.80     762.71         7.50  105            1
1    Tiffany     Myers        Music therapist    201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19          20        2023  XDA18116-89-4A  consumables                pens       24.95      15.09         1.49   21            0
2    Shannon    Watson           Immunologist   687.737.9424x8503                swatson8146@payne.net      57113   1902144  2023-01-29           4        2023  ABE59463-05-7E     software  project management       13.52       7.86         1.68   39            6
3     Nathan     Baker  Accounting technician    827-788-8123x012               bakernathan@benson.com      46554   9031802  2023-04-25          17        2023  ZMM00836-65-0C  consumables                pens       36.42      24.85         1.23   29            3
4  Christina  Schwartz           Chiropractor        265-829-3643  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28          21        2023  BZX55559-12-3X  consumables                misc      195.10     108.17        46.43   20            1

In [31]:
# I need these codes to show the result without wrapping and not limit the width
pd.set_option('display.max_columns', None)  # Show all columns
pd.set_option('display.width', None)        # Don't limit the width
pd.set_option('display.max_rows', None)     # Show all rows
pd.set_option('display.expand_frame_repr', False)  # Don't wrap to multiple lines


In [32]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [33]:
# Use the describe function to gather some basic statistics
df.describe() #describe the data don't forget the () at the end

client_id      order_id    order_week    order_year    unit_price     unit_cost   unit_weight           qty   line_number
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000  54639.000000  54639.000000  5.463900e+04  54639.000000
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207     99.446073      5.004116  5.702646e+02      2.979667
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135    133.164267      5.326599  1.879552e+04      2.436320
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000      0.010000      0.000000  0.000000e+00      0.000000
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000     14.840000      1.450000  3.200000e+01      1.000000
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000     49.890000      3.240000  6.800000e+01      3.000000
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000    125.570000      6.890000  1.700000e+02      5.000000
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000    846.270000     46.430000  3.958244e+06      9.000000

In [34]:
# Use the info function to gather some basic information
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info() #info about the data type. Ie: object(aka string/mixed types), integer, float, bool, etc


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [35]:
# What three item categories had the most entries?
# Get counts of each category in the 'category' column and sort descending then do head(3) to give me the top 3
category_counts = df['category'].value_counts().sort_values(ascending=False)
category_counts.head(3)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [36]:
#### EXTRA ####
# Get the max count for the top 3 category then give me the highest count from its subcategory 

category_top3_df = df[df['category'].isin(category_counts.head(3).index)]

max_counts = (category_top3_df.groupby(['category', 'subcategory']) #takes the DF and groups it by both category and subcat together
              .size() # counts how many times each category-subcategory pair appears
              .reset_index(name='count') #converts the grouped data back into a regular DF, name the count column  = 'count'
              .sort_values('count', ascending=False)
              .groupby('category') # group by category
              .first())


max_counts

subcategory  count
category                             
consumables  bathroom supplies   6424
furniture               tables   4540
software            accounting   3756

In [37]:
# For the category with the most entries, which subcategory had the most entries?
# this gives you the bathroom supplies subcategory from the consumables category
consumables_subcats = df[df['category'] == 'consumables']['subcategory'].value_counts()
print(consumables_subcats.head(1))

subcategory
bathroom supplies    6424
Name: count, dtype: int64


In [38]:
# Which five clients had the most entries in the data?
df['client_id'].value_counts().sort_values(ascending=False).head(5)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [39]:
# Store the client ids of those top 5 clients in a list.
top_five_clients_df = df['client_id'].value_counts().sort_values(ascending=False).head(5).index.tolist()

top_five_clients_df


[33615, 66037, 46820, 38378, 24741]

In [40]:
# How many total units (the qty column) did the client with the most entries order?

df[df['client_id'] == 33615]['qty'].sum()

64313

In [41]:
#### EXTRA ####
# Better way to structure this without doing the above, if you don't want to put client 33615, this way is more streamlined
top_clients = df['client_id'].value_counts().head()
print("\nTop 5 clients by number of entries:")
print(top_clients)

top_client_ids = top_clients.index.tolist()[:5]
print("\nTop 5 client IDs:")
print(top_client_ids)

top_client_units = df[df['client_id'] == top_client_ids[0]]['qty'].sum()
print(f"\nTotal units ordered by client {top_client_ids[0]}:")
print(top_client_units)


Top 5 clients by number of entries:
client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

Top 5 client IDs:
[33615, 66037, 46820, 38378, 24741]

Total units ordered by client 33615:
64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [42]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
# first way is combining a print statement in one row, where you define what the math logic is then call out the rest
# print(df.assign(subtotal=df['unit_price'] * df['qty'])[['unit_price', 'qty', 'subtotal']].head(2))

# or you can use another approach where you define what subtotal logic is then call out all the columns
# both ways give the same answer
df['subtotal'] = df['unit_price'] * df['qty']
print(df[['unit_price', 'qty', 'subtotal']].head(2))    



   unit_price  qty   subtotal
0     1096.80  105  115164.00
1       24.95   21     523.95


In [43]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.

# first I need to get the total weight calculation
df['total_weight'] = df['unit_weight'] * df['qty']

# then put the conditional logic 
df['shipping_price'] = round(df['total_weight'].apply(lambda x: 7 if x > 50 else 10) * df['total_weight'],2)

# show me the columns as result of the above logic
#print(df[['unit_price', 'unit_weight', 'qty', 'total_weight', 'shipping_price']])

# just give the top 3 rows
print(df[['unit_price', 'unit_weight', 'qty', 'total_weight', 'shipping_price']].head(3))

   unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64


In [44]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['line_price'] = round((df['subtotal'] + df['shipping_price'])*1.0925,2)

# show me the columns as result of the above logic
#print(df[['subtotal', 'shipping_price', 'line_price']])

# just give the top 3 rows to show results
print(df[['subtotal', 'shipping_price', 'line_price']].head(3))

    subtotal  shipping_price  line_price
0  115164.00         5512.50   131839.08
1     523.95          312.90      914.26
2     527.28          458.64     1077.12


In [57]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).

df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']

print(df[['first', 'last', 'job', 'phone', 'email', 'client_id',
          'order_date', 'order_week', 'order_year', 'unit_price', 
          'unit_cost', 'unit_weight', 'qty', 'line_number', 'subtotal', 
          'total_weight', 'shipping_price', 'line_price', 'line_cost']].head(3))

     first     last              job               phone                              email  client_id  order_date  order_week  order_year  unit_price  unit_cost  unit_weight  qty  line_number   subtotal  total_weight  shipping_price  line_price  line_cost
0   Donald  Harding     Immunologist  793-904-7725x39308   harding.donald.7185@sullivan.com      58515  2023-04-28          17        2023     1096.80     762.71         7.50  105            1  115164.00        787.50         5512.50   131839.08   85597.05
1  Tiffany    Myers  Music therapist    201.442.4543x942  myers.t.6537@ferguson-johnson.net      37609  2023-05-19          20        2023       24.95      15.09         1.49   21            0     523.95         31.29          312.90      914.26     629.79
2  Shannon   Watson     Immunologist   687.737.9424x8503              swatson8146@payne.net      57113  2023-01-29           4        2023       13.52       7.86         1.68   39            6     527.28         65.52          45

In [47]:
# Create a column for the profit of each line using line cost and line price
df['line_profit'] = df['line_price'] - df['line_cost']

#These set_option are already defined at the very top of the code, so you don't need to redefine them. But if you haven't make sure you do the two lines of codes below.
#pd.set_option('display.max_columns', None)
#pd.set_option('display.width', None)

print(df.head(3))

     first     last              job               phone                              email  client_id  order_id  order_date  order_week  order_year         item_id     category         subcategory  unit_price  unit_cost  unit_weight  qty  line_number   subtotal  total_weight  shipping_price  line_price  line_cost  line_profit
0   Donald  Harding     Immunologist  793-904-7725x39308   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28          17        2023  EUD29711-63-6U        decor            wall art     1096.80     762.71         7.50  105            1  115164.00        787.50         5512.50   131839.08   85597.05     46242.03
1  Tiffany    Myers  Music therapist    201.442.4543x942  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19          20        2023  XDA18116-89-4A  consumables                pens       24.95      15.09         1.49   21            0     523.95         31.29          312.90      914.26     629.79       284.47
2  Shannon   

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [48]:
# Check your work against the totals above

"""
# Test with a simpler code:
# I want to see the details first and then the total for each of the order id 2742071:
print(df[df['order_id'] == 2742071][['order_id', 'line_price']])
print(f"Order ID {2742071} had a total price of ${df[df['order_id'] == 2742071]['line_price'].sum():,.2f}")
print("------------------------------------------------------------------------------------------------------------------------")
"""

# Define the order IDs and columns we want to see
order_ids = [2742071, 2173913, 6128929]
columns_to_show = ['order_id', 'line_price']

# Loop through each order ID
for order_id in order_ids:
    # Print the details
    print(df[df['order_id'] == order_id][columns_to_show])
    
    # Print the total
    total = df[df['order_id'] == order_id]['line_price'].sum()
    print(f"Order ID {order_id} had a total price of ${total:,.2f}")
    
    # Print separator
    print("-" * 100)

       order_id  line_price
3446    2742071     6629.69
6287    2742071   135889.52
20623   2742071     2069.02
23840   2742071     1273.61
24671   2742071     1976.73
42470   2742071     4973.32
Order ID 2742071 had a total price of $152,811.89
----------------------------------------------------------------------------------------------------
       order_id  line_price
627     2173913     9856.86
14089   2173913      538.03
14938   2173913     7092.74
14962   2173913     5076.03
22004   2173913     9640.08
27713   2173913     4297.41
53155   2173913   125887.56
Order ID 2173913 had a total price of $162,388.71
----------------------------------------------------------------------------------------------------
       order_id  line_price
3422    6128929    48763.65
13672   6128929     8121.21
14517   6128929   180327.76
16295   6128929     4133.06
21141   6128929     1713.39
37265   6128929     2144.96
37949   6128929   367345.26
38542   6128929      173.84
48834   6128929   102182.6

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [58]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.

"""
# Test with a simpler code:
# I want to see the details first and then the total for each of the client id 33615:
print(df[df['client_id'] == 33615][['client_id', 'line_price']])
print(f"Client ID {33615} had a total price of ${df[df['client_id'] == 33615]['line_price'].sum():,.2f}")
print("------------------------------------------------------------------------------------------------------------------------")
"""

# Define the client IDs and columns we want to see
top_client_ids
columns_to_show = ['client_id', 'line_price']

# Loop through each client ID the sum
for client_id in top_client_ids:
    # Print the details if necessary to check
    # print(df[df['client_id'] == client_id][columns_to_show])
    
    # Print the total
    total = df[df['client_id'] == client_id]['line_price'].sum()
    print(f"Client ID {client_id} had a total price of ${total:,.2f}")
    
    # Print separator
    print("-" * 100)

Client ID 33615 had a total price of $8,377,308.52
----------------------------------------------------------------------------------------------------
Client ID 66037 had a total price of $10,259,514.79
----------------------------------------------------------------------------------------------------
Client ID 46820 had a total price of $9,743,794.36
----------------------------------------------------------------------------------------------------
Client ID 38378 had a total price of $12,906,550.87
----------------------------------------------------------------------------------------------------
Client ID 24741 had a total price of $82,268,892.02
----------------------------------------------------------------------------------------------------


In [52]:
#### EXTRA ####
## OR use this group by approach, the order is a bit different than above but same results
print(top_client_ids)

client_totals = df[df['client_id'].isin(top_client_ids)].groupby('client_id')['line_price'].sum()
for client_id, total in client_totals.items():
    print(f"Total for Client {client_id}: ${total:,.2f}")
    print("-" * 100)

[33615, 66037, 46820, 38378, 24741]
Total for Client 24741: $82,268,892.02
----------------------------------------------------------------------------------------------------
Total for Client 33615: $8,377,308.52
----------------------------------------------------------------------------------------------------
Total for Client 38378: $12,906,550.87
----------------------------------------------------------------------------------------------------
Total for Client 46820: $9,743,794.36
----------------------------------------------------------------------------------------------------
Total for Client 66037: $10,259,514.79
----------------------------------------------------------------------------------------------------


Instructions:
* Format the data and rename the columns to names suitable for presentation.

* Define the money columns. 

* Define a function that converts a dollar amount to millions.

* Apply the currency_format_millions function to only the money columns. 

* Rename the columns to reflect the change in the money format. 

In [59]:
# Create a summary DataFrame showing the totals for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 

summary_data = []
for client_id in top_client_ids:
    client_df = df[df['client_id'] == client_id]
    
    client_summary = {
        'client_id': client_id,
        'qty': client_df['qty'].sum(),
        'shipping_price': client_df['shipping_price'].sum(),
        'line_price': client_df['line_price'].sum(),
        'line_cost': client_df['line_cost'].sum(),
        'line_profit': client_df['line_profit'].sum()
    }
    summary_data.append(client_summary)

# Create Data Frame from summary_data above
summary_df = pd.DataFrame(summary_data)

print(summary_df)

   client_id     qty  shipping_price   line_price    line_cost  line_profit
0      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61
1      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81
2      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
3      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
4      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31


In [60]:
#### this section relies heavily from the section above it, otherwise you will get an error

# Convert money values to millions using a function
money_columns = ['shipping_price', 'line_price', 'line_cost', 'line_profit']
for column in money_columns:
    summary_df[column] = summary_df[column] / 1000000
    

# Rename columns to make it suitable for presentation
new_column_names = {
    'client_id': 'Client ID',
    'qty': 'Units',
    'shipping_price': 'Shipping (millions)',
    'line_price': 'Total Revenue (millions)',
    'line_cost': 'Total Cost (millions)',
    'line_profit': 'Total Profit (millions)'
}
summary_df = summary_df.rename(columns=new_column_names)

# I want to add dollar formatting to the renamed columns
formatted_money_columns = ['Shipping (millions)', 'Total Revenue (millions)', 
                         'Total Cost (millions)', 'Total Profit (millions)']
for column in formatted_money_columns:
    summary_df[column] = summary_df[column].map('${:,.6f}'.format)

print(summary_df)

   Client ID   Units Shipping (millions) Total Revenue (millions) Total Cost (millions) Total Profit (millions)
0      33615   64313           $1.828985                $8.377309             $6.175314               $2.201995
1      66037   43018           $1.395152               $10.259515             $7.004483               $3.255032
2      46820   75768           $1.601449                $9.743794             $7.007192               $2.736603
3      38378   73667           $3.429455               $12.906551             $9.634721               $3.271830
4      24741  239862           $5.126448               $82.268892            $45.688900              $36.579992


In [61]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DataFrame.
sorted_by_total_profit_df = summary_df.sort_values('Total Profit (millions)', ascending=False) #ascending = False means it's descending order
print("\nFinal sorted summary:")
print(sorted_by_total_profit_df)


Final sorted summary:
   Client ID   Units Shipping (millions) Total Revenue (millions) Total Cost (millions) Total Profit (millions)
4      24741  239862           $5.126448               $82.268892            $45.688900              $36.579992
3      38378   73667           $3.429455               $12.906551             $9.634721               $3.271830
1      66037   43018           $1.395152               $10.259515             $7.004483               $3.255032
2      46820   75768           $1.601449                $9.743794             $7.007192               $2.736603
0      33615   64313           $1.828985                $8.377309             $6.175314               $2.201995


# SUMMARY 
* On the last part above, I wanted to show that the values in all columns with suffix (millions) are reflected in $ amounts, hence the $ sign
* Client ID 24741 has the highest total profit $36.579992 Millions
* Client ID 33615 has the lowest total profit $2.201995 Millions
* The final sorted summary report did not have the index resetted, so if we need to reset it so it would show consequential 0,1,2,3,4 then we need to add the following statement
sorted_by_total_profit_df = sorted_by_total_profit_df.reset_index(drop=True)
* See example below

In [62]:
# Make the index shows consequential, same result for other columns as above
sorted_by_total_profit_df = summary_df.sort_values('Total Profit (millions)', ascending=False)
sorted_by_total_profit_df = sorted_by_total_profit_df.reset_index(drop=True)
print("\nFinal sorted summary:")
print(sorted_by_total_profit_df)


Final sorted summary:
   Client ID   Units Shipping (millions) Total Revenue (millions) Total Cost (millions) Total Profit (millions)
0      24741  239862           $5.126448               $82.268892            $45.688900              $36.579992
1      38378   73667           $3.429455               $12.906551             $9.634721               $3.271830
2      66037   43018           $1.395152               $10.259515             $7.004483               $3.255032
3      46820   75768           $1.601449                $9.743794             $7.007192               $2.736603
4      33615   64313           $1.828985                $8.377309             $6.175314               $2.201995
